crates/sled/src/tree.rs

```diff
@@ -288,6 +288,41 @@ impl Tree {
         }
     }
 
+    /// fetch the value, apply a function to it and return the result
+    pub fn update_and_fetch<K, F>(&self, key: K, mut fun: F) -> Result<Option<Vec<u8>>>
+    where K: AsRef<[u8]>,
+          F: FnMut(Option<&[u8]>) -> Option<Vec<u8>>,
+    {
+        let key = key.as_ref();
+        let mut current = self.get(key)?;
+
+        loop {
+            let tmp = current.as_ref().map(AsRef::as_ref);
+            let next = fun(tmp);
+            match self.cas(key, tmp, next.clone())? {
+                Ok(()) => return Ok(next),
+                Err(new_current) => current = new_current,
+            }
+        }
+    }
+
+    /// fetch the value, apply a function to it and return the previous value
+    pub fn fetch_and_update<K, F>(&self, key: K, mut fun: F) -> Result<Option<Vec<u8>>>
+    where K: AsRef<[u8]>,
+          F: FnMut(Option<&[u8]>) -> Option<Vec<u8>>,
+    {
+        let key = key.as_ref();
+        let mut current = self.get(key)?;
+
+        loop {
+            let tmp = current.as_ref().map(AsRef::as_ref);
+            match self.cas(key, tmp, fun(tmp))? {
+                Ok(()) => return Ok(current.map(|v| v.to_vec())),
```

Comment:
this should return an IVec instead of copying the previous value to create the Vec<u8>